File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
from collections import Counter


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        s_len, p_len = len(s), len(p)
        if s_len < p_len:
            return []
        p_counter = Counter(p)
        word_counter = Counter(s[:p_len])
        result = [0] if word_counter == p_counter else []
        for i in range(s_len - p_len):
            word_counter[s[i]] -= 1
            word_counter[s[i + p_len]] += 1
            if word_counter == p_counter:
                result.append(i + 1)
        return result

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int s_size = s.size();
        int p_size = p.size();

        if (s_size < p_size) {
            return {};
        }

        vector<int> result;
        vector<int> target(26, 0);
        vector<int> window(26, 0);

        for (int i = 0; i < p_size; ++i) {
            target[p[i] - 'a'] += 1;
            window[s[i] - 'a'] += 1;
        }
        if (window == target) {
            result.push_back(0);
        }

        for (int i = p_size; i < s_size; ++i) {
            window[s[i] - 'a'] += 1;
            window[s[i - p_size] - 'a'] -= 1;
            if (window == target) {
                result.push_back(i - p_size + 1);
            }
        }
        return result;
    }
};
